From b0bd3145a432a2cc347a35b1872889c317967b49 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Tue, 20 Jan 2015 17:17:00 -0500
Subject: [fs] cifs: Complete oplock break jobs before closing file handle

Message-id: <1421774220-7270-1-git-send-email-sprabhu@redhat.com>
Patchwork-id: 103209
O-Subject: [RHEL7 PATCH fs] [bz 1177215]Complete oplock break jobs before closing file handle
Bugzilla: 1177215
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

Original summary follows
--
Commit
c11f1df5003d534fd067f0168bfad7befffb3b5c
requires writers to wait for any pending oplock break handler to
complete before proceeding to write. This is done by waiting on bit
CIFS_INODE_PENDING_OPLOCK_BREAK in cifsFileInfo->flags. This bit is
cleared by the oplock break handler job queued on the workqueue once it
has completed handling the oplock break allowing writers to proceed with
writing to the file.

While testing, it was noticed that the filehandle could be closed while
there is a pending oplock break which results in the oplock break
handler on the cifsiod workqueue being cancelled before it has had a
chance to execute and clear the CIFS_INODE_PENDING_OPLOCK_BREAK bit.
Any subsequent attempt to write to this file hangs waiting for the
CIFS_INODE_PENDING_OPLOCK_BREAK bit to be cleared.

We fix this by ensuring that we also clear the bit
CIFS_INODE_PENDING_OPLOCK_BREAK when we remove the oplock break handler
from the workqueue.

The bug was found by Red Hat QA while testing using ltp's fsstress
command.

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
Signed-off-by: Jeff Layton <jlayton@samba.org>
(cherry picked from commit a3c35f5c8cd0439a3bda30e5d387160a60529aab)
--

This was successfully tested by running the fsstress command in a loop.
The patch is currently in Steve French's git tree for submission into
Linus's tree.

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/file.c b/fs/cifs/file.c
index 12f9ac5..6201bf3 100644
--- a/fs/cifs/file.c
+++ b/fs/cifs/file.c
@@ -366,6 +366,7 @@ void cifsFileInfo_put(struct cifsFileInfo *cifs_file)
  struct cifsLockInfo *li, *tmp;
  struct cifs_fid fid;
  struct cifs_pending_open open;
+ bool oplock_break_cancelled;
 
  spin_lock(&cifs_file_list_lock);
  if (--cifs_file->count > 0) {
@@ -397,7 +398,7 @@ void cifsFileInfo_put(struct cifsFileInfo *cifs_file)
  }
  spin_unlock(&cifs_file_list_lock);
 
- cancel_work_sync(&cifs_file->oplock_break);
+ oplock_break_cancelled = cancel_work_sync(&cifs_file->oplock_break);
 
  if (!tcon->need_reconnect && !cifs_file->invalidHandle) {
   struct TCP_Server_Info *server = tcon->ses->server;
@@ -409,6 +410,9 @@ void cifsFileInfo_put(struct cifsFileInfo *cifs_file)
   _free_xid(xid);
  }
 
+ if (oplock_break_cancelled)
+  cifs_done_oplock_break(cifsi);
+
  cifs_del_pending_open(&open);
 
  /*
-- 
1.7.1